sygnal/firebasepushkin.py

```diff
@@ -0,0 +1,263 @@
+    "sygnal_fcm_request_time", "Time taken to send HTTP request"
+)
+
+MAX_TRIES = 3
+RETRY_DELAY_BASE = 10
+MAX_BYTES_PER_FIELD = 1024
+DEFAULT_MAX_CONNECTIONS = 20
+
+logger = logging.getLogger(__name__)
+
+
+@attr.s
+class FirebaseConfig(object):
+    credentials = attr.ib()
+    max_connections = attr.ib(default=20)
+    message_types = attr.ib(default=attr.Factory(dict), type=Dict[str, str])
+
+
+class FirebasePushkin(Pushkin):
+    def __init__(self, name, sygnal, config):
+        super(FirebasePushkin, self).__init__(name, sygnal, config)
+
+        self.db = sygnal.database
+        self.reactor = sygnal.reactor
+        self.config = FirebaseConfig(
+            **{x: y for x, y in self.cfg.items() if x != "type"}
+        )
+
+        credential_path = self.config.credentials
+        if not credential_path:
+            raise PushkinSetupException("No Credential path set in config")
+
+        cred = credentials.Certificate(credential_path)
+
+        self._pool = ThreadPool(maxthreads=self.config.max_connections)
+        self._pool.start()
+
+        self._app = initialize_app(cred, name="app")
+
+    async def dispatch_notification(self, n, device, context):
+
+        pushkeys = [
+            device.pushkey for device in n.devices if device.app_id == self.name
+        ]
+
+        failed = []
+        data = self.build_message(n)
+
+        logger.debug("Type: %s", data["type"])
+
+        if (
+            data["type"] != "m.room.message"
+            or data["content"]["msgtype"] not in self.config.message_types
+        ):
+            return failed
+
+        unread_count = data["unread"] if data["unread"] is not None else 0
+        notification_title = (
+            data["sender_display_name"]
+            if data["room_name"] is None
+            else data["room_name"]
+        )
+
+        if data["content"]["msgtype"] == "m.text":
+            message = self.text_message_notification(
+                data, notification_title, unread_count, pushkeys
+            )
+        else:
+            message = self.default_notification(
+                data,
+                notification_title,
+                unread_count,
+                pushkeys,
+                self.config.message_types[data["content"]["msgtype"]],
+            )
+
+        failed.extend(await self.send(message))
+
+        return failed
+
+    async def send(self, message):
+
+        d = Deferred()
+
+        def done(success, result):
+            self.reactor.callFromThread(d.callback, result)
+
+        with SEND_TIME_HISTOGRAM.time():
+            self._pool.callInThreadWithCallback(
+                done, messaging.send_multicast, message, app=self._app
+            )
+            response = await d
+
+        logger.debug(
+            "Message send success: %s of %s",
+            response.success_count,
+            response.success_count + response.failure_count,
+        )
+
+        failed = []
+
+        return failed
```

Comment:
Did we want to do something here rather than reporting success for everything? Did we want to catch exceptions from the send?